File: spark_java/src/main/java/com/spark/study/topn/GroupTop3.java
package com.spark.study.topn;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.PairFunction;
import org.apache.spark.api.java.function.VoidFunction;
import scala.Tuple2;

import java.util.Arrays;
import java.util.Iterator;

/**
 * @author yangqian
 * @date 2021/5/9
 */
public class GroupTop3 {

    public static void main(String[] args) {
        SparkConf conf = new SparkConf().setAppName("GroupTop3").setMaster("local");
        final JavaSparkContext sc = new JavaSparkContext(conf);
        JavaRDD<String> lines = sc.textFile("/Users/qian/WorkSpaces/own-workspace/2021/spark_code_repository/spark_java/src/main/resources/score.txt");
        JavaPairRDD<String, Integer> scoreWithClass = lines.mapToPair(new PairFunction<String, String, Integer>() {
            private static final long serialVersionUID = 5927267795933596641L;

            @Override
            public Tuple2<String, Integer> call(String line) throws Exception {
                return new Tuple2<String, Integer>(line.split(" ")[0], Integer.parseInt(line.split(" ")[1]));
            }
        });
        // 按照班分组数据（分数）
        JavaPairRDD<String, Iterable<Integer>> groupPairs = scoreWithClass.groupByKey();

        JavaPairRDD<String, Iterable<Integer>> top3Score = groupPairs.mapToPair(new PairFunction<Tuple2<String, Iterable<Integer>>, String, Iterable<Integer>>() {
            private static final long serialVersionUID = -3928427695844412087L;

            @Override
            public Tuple2<String, Iterable<Integer>> call(
                    Tuple2<String, Iterable<Integer>> classScores) throws Exception {
                String className = classScores._1;
                Iterator<Integer> scores = classScores._2.iterator();
                Integer[] top3 = new Integer[3];
                while (scores.hasNext()) {
                    Integer score = scores.next();
                    for (int i = 0; i < 3; i++) {
                        if (top3[i] == null) {
                            top3[i] = score;
                        } else if (score > top3[i]) {
                            int tmp = top3[i];
                            top3[i] = score;

                            if (i < top3.length - 1) {
                                top3[i + 1] = tmp;
                            }
                        }
                    }
                }
                return new Tuple2<String, Iterable<Integer>>(className, Arrays.asList(top3));
            }
        });

        top3Score.foreach(new VoidFunction<Tuple2<String, Iterable<Integer>>>() {
            private static final long serialVersionUID = -7251855902775553443L;

            @Override
            public void call(Tuple2<String, Iterable<Integer>> tuple2) throws Exception {
                System.out.println(tuple2._1 + "  --  " + tuple2._2);
            }
        });

    }

}

File: spark_scala/src/main/scala/com/spark/scala/lesson/topn/GroupTop3.scala
package com.spark.scala.lesson.topn

import org.apache.spark.{SparkConf, SparkContext}

/**
 * @date 2021/5/10
 */
object GroupTop3 {

  def main(args: Array[String]): Unit = {
    val conf = new SparkConf().setAppName("groupTop3").setMaster("local")
    val sc = new SparkContext(conf)

    val lines = sc.textFile("/Users/qian/WorkSpaces/own-workspace/2021/spark_code_repository/spark_java/src/main/resources/score.txt")
    val groupClassScoreRdd = lines.map(line => (line.split(" ")(0), line.split(" ")(1)))
    val groupClassScoreListRdd = groupClassScoreRdd.groupByKey()
    val sortClassScoreListRdd = groupClassScoreListRdd.map(classScore => {
      val className = classScore._1
      val scores = classScore._2.iterator
      val top3 = new Array[Int](3)
      var flag = true
      while (scores.hasNext) {
        val score = scores.next
        for (i <- 0 until 3 if flag) {
          if (top3(i) == 0) {
            top3(i) = score.toInt
          } else if (score.toInt > top3(i)) {
            val tmp = top3(i)
            top3(i) = score.toInt
            if (i < top3.length - 1) {
              top3(i + 1) = tmp
              flag = false
            }
          }
        }
      }
      (className, top3)
    })
    sortClassScoreListRdd.foreach(r => println(s"className:${r._1}, score Top3:${r._2.mkString("Array(", ", ", ")")}"))
  }

}
